### PR TITLE
C# extractor: Limit trap strings to 1MB

### PR DESCRIPTION
**Pull request by calumgrant**
This is made more complicated by the fact that we need to limit the size of the encoded output to 1MB, so we can't just truncate each string to 1M characters.

There are 2 test cases:

1. When the string consists of ASCII, 1-byte UTF-8 characters, in which case we get strings up to 1M characters (1048576 characters).
2. When the string consists of multi-byte UTF8 characters. In this case, we can only store less than 1M characters, in this case 443093.

### CODE REVIEW

File: csharp/extractor/Semmle.Extraction/Tuple.cs [hunk 1]
```diff
@@ -15,6 +17,68 @@ public Tuple(string name, params object[] args)
...
+            return encoding.GetMaxByteCount(s.Length) > maxStringBytes &&
+                encoding.GetByteCount(s) > maxStringBytes;
+        }
+
+        private static bool NeedsTruncation(string[] array)
+        {
+            // Optimization: only count the actual number of bytes if there is the possibility 
+            // of the strings exceeding maxStringBytes
+            return encoding.GetMaxByteCount(array.Sum(s => s.Length)) > maxStringBytes &&
+                array.Sum(s => encoding.GetByteCount(s)) > maxStringBytes;
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction/Tuple.cs]: `array.Sum(encoding.GetByteCount)`

File: csharp/extractor/Semmle.Extraction/Tuple.cs [hunk 2]
```diff
@@ -27,48 +91,65 @@ public void EmitToTrapBuilder(ITrapBuilder tb)
             foreach (var a in Args)
             {
                 if (column > 0) tb.Append(", ");
-                if (a is Label)
+                if (a is Label l)
```
[the remark this answers is not in the document]
Comment (calumgrant) [reply csharp/extractor/Semmle.Extraction/Tuple.cs]: Sounds like a good idea for a query.

File: csharp/extractor/Semmle.Extraction/Tuple.cs (hunk 1 above)
Comment (calumgrant) [reply csharp/extractor/Semmle.Extraction/Tuple.cs]: Sounds like a good idea for a query.

**Comment by pavgust** [comment]
This pull request **introduces 1 alert** when merging e18eeb8d2a19a622fc65e7362a0612de68ffa6c0 into 97fc4b0c8e254f9661eeee6ae0078e96ba818af6 - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-83197962483cf534ac4e884de1b6458589fd9d55)

**new alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Pull request merged**